File: c++/balanced_binary_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

#include <iostream>
using std::cout;
using std::endl;

struct TreeNode
{
	int val;
	TreeNode* left;
	TreeNode* right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

typedef struct TreeNode* Tree;

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        int height;
        return isTreeBalanced(root, height);
    }
private:
		bool isTreeBalanced(TreeNode* root, int& height)
		{
			bool leftBalanced, rightBalanced, rootBalanced;
			int leftHeight, rightHeight;

			if (root == NULL)
			{
				height = 0;
				return true;
			}

			if (root -> left == NULL && root -> right == NULL)
			{
				height = 1;
				return true;
			}

			leftBalanced = isTreeBalanced(root -> left, leftHeight);
			rightBalanced = isTreeBalanced(root -> right, rightHeight);
			rootBalanced = abs(leftHeight - rightHeight) > 1 ? false : true;

			height = max(leftHeight, rightHeight) + 1;

			return leftBalanced && rightBalanced && rootBalanced;
		}

		int max(int a, int b)
		{
			return a > b ? a : b;
		}

		int abs(int a)
		{
			if (a < 0)
				return a * -1;

			return a;
		}
};

Tree Insert(int value, Tree T)
{
	if (T == NULL)
		return new TreeNode(value);

	if (T -> val > value)
		T -> left = Insert(value, T -> left);
	else if (T -> val < value)
		T -> right = Insert(value, T -> right);

	return T;
}

void PostOrderTraverse(Tree T)
{
	if (T == NULL)
		return;

	PostOrderTraverse(T -> left);
	PostOrderTraverse(T -> right);
	cout << T -> val << " ";
}


void InOrderTraverse(Tree T)
{
	if (T == NULL)
		return;

	InOrderTraverse(T -> left);
	cout << T -> val << " ";
	InOrderTraverse(T -> right);
}

void PreOrderTraverse(Tree T)
{
	if (T == NULL)
		return;

	cout << T -> val << " ";
	PreOrderTraverse(T -> left);
	PreOrderTraverse(T -> right);
}

int main(int argc, char const *argv[])
{
	int array[] = {7, 3, 8};
	Tree T;
	int i, Len, balanced;
	Solution s;

	Len = sizeof(array) / sizeof(array[0]);
	T = NULL;
	for (i = 0; i < Len; ++i)
		T = Insert(array[i], T);

	PreOrderTraverse(T);
	cout << endl;

	InOrderTraverse(T);
	cout << endl;

	PostOrderTraverse(T);
	cout << endl;

	balanced = s.isBalanced(T);
	cout << balanced << endl;
	return 0;
}

File: javascript/balanced_binary_tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
function TreeNode(val) {
	 this.val = val;
   this.left = this.right = null;
}

var abs = function(a) {
	return a > 0 ? a : -1 * a;
}

var max = function(a, b) {
	return a > b ? a : b;
}

var isTreeBalanced = function(root, height) {
	var leftBalanced, rightBalanced;
	var leftHeight, rightHeight;

	if (root === null)
	{
		height.value = 0;
		return true;
	}

	if (root.left === null && root.right === null)
	{
		height.value = 1;
		return true;
	}

	leftHeight = {};
	rightHeight = {};
	leftBalanced = isTreeBalanced(root.left, leftHeight);	
	rightBalanced = isTreeBalanced(root.right, rightHeight);

	height.value = max(leftHeight.value, rightHeight.value) + 1;
	return abs(leftHeight.value - rightHeight.value) <= 1 && leftBalanced && rightBalanced;
};

var isBalanced = function(root) {
   var height = {};
   return isTreeBalanced(root, height);
};

var insert = function(root, value) {
	if (root == null)
		return new TreeNode(value);

	if (root.val > value)
		root.left = insert(root.left, value);
	else if (root.val < value)
		root.right = insert(root.right, value);

	return root;
}

var postTraverse = function(root) {
	if (root == null)
		return;

	console.log(root.val);
	postTraverse(root.left);
	postTraverse(root.right);
}

var array = [1, 2, 3]
var i;
var root;

root = null
for (i = 0; i < array.length; ++i) {
	root = insert(root, array[i]);
}

console.log(isBalanced(root));

